22/2.rs

```diff
@@ -0,0 +1,267 @@
+use std::fs;
+use std::cmp::max;
+use std::cmp::min;
+use std::convert::TryInto;
+
+type ParseTarget = Vec<Command>;
+type Solution = usize;
+
+const EXAMPLES: [(&str, Solution); 3] = [
+    ("1", 39),
+    ("2", 39769202357779),
+    ("3", 2758514936282235),
+];
+
+const DAY: u8 = 22;
+
+fn main() {
+    let results = EXAMPLES.iter()
+        .zip(
+            EXAMPLES.iter()
+                .map(|t| format!("{}/{}.ie", DAY, t.0))
+                .map(operation)
+        )
+        .map(|((name, expected), result)| 
+            (
+                *name,
+                result
+                    .and_then(|actual| if *expected == actual {
+                        return Ok(());
+                    } else {
+                        return Err(format!("Expected {} but got {}", expected, actual));
+                    })
+            )
+        )
+        .collect::<Vec<(&str, Result<(), String>)>>();
+    results.iter()
+        .for_each(|(name, result)| match result {
+            Ok(()) => println!("Example {} passed.", name),
+            Err(msg) => println!("Example {} failed: {}.", name, msg)
+        });
+
+    if results.iter().any(|t| t.1.is_err()) {
+        panic!("Please address errors before attempting the problem.")
+    }
+
+    println!(
+        "{}",
+        operation(format!("{}/input", DAY)).expect("Unexpected Error in main input.")
+    );
+}
+
+// fn error<T>(msg: &str) -> Result<T, String> {
+//     return Err(String::from(msg));
+// }
+
+fn operation(filename: String) -> Result<Solution, String> {
+    return fs::read_to_string(filename)
+        .map_err(|io_error| format!("{}", io_error))
+        .and_then(parse)
+        .and_then(solve);
+}
+
+#[derive(PartialEq, Eq)]
+struct Command {
+    on: bool,
+    points: Points
+}
+
+impl Command {
+    fn from_input(line: &str) -> Result<Command, String> {
```

Comment:
So, using `ok_or().and_then()` like this is great so long as nothing goes wrong. The sticky wicket is, your `ok_or` is going to unwrap your `Result`, so you'll be passing strings you can't parse into your `and_then` function. 

You've got a lot of options, if you want them. The smallest, simplest change would be to use `map`:
```
bit.next().map(|s| parse_range("x=", s)).ok_or( /* etc */)
```
But you can capture that whole thing into `map_or`:
```
bit.next().map_or("insufficient segments for x range", |s| parse_range("x=", s)); 
```

Or, you could go wild, and explore some other ways of getting DRY-er, if you want: 
```
let point_bits: Vec<&str> = bits.take(3).collect();
// Assert on the length of point_bits if desired, it's AoC so you do you 
// Now some options:
// Here, these three things should already be unwrapped Ranges
let (x_range, y_range, z_range) =  point_bits.zip(["x=", "y=", "z="]).filter_map(|bit, prefix| parse_range(prefix, bit)).collect::<(Range, Range, Range)>();
// Or:
let worked: Result<Vec<Range>, String> =  point_bits.zip(["x=", "y=", "z="]).map(|bit, prefix| parse_range(prefix, bit)).collect();
// Allows directly destructuring your Ok vec
match worked {
    Ok([x, y, z]) => Ok(Command{on, points: Point {xs, ys, zs}}),
    Error(bad) => Error(bad), // this part isn't as nice
}
```

Reply:
Collect did _NOT_ want to collect down to a tuple.  Not sure why.